stdlib/public/core/MutableCollection.swift

```diff
@@ -194,29 +194,46 @@ where SubSequence: MutableCollection
   ) rethrows -> R?
 
-  /// Call `body(buffer)`, where `buffer` provides access to the contiguous
-  /// mutable storage of the entire collection. If no such storage exists, it is
-  /// first created. If the collection does not support an internal
-  /// representation in the form of contiguous mutable storage, `body` is not
-  /// called and `nil` is returned.
+  /// Executes a closure on the collection's contiguous storage.
+  ///
+  /// This method calls `body(buffer)`, where `buffer` provides access to the
+  /// contiguous mutable storage of the entire collection. If no such storage
+  /// exists, the collection creates is. If the collection doesn't support an
+  /// internal representation in the form of contiguous mutable storage, this
+  /// method doesn't call `body` and immediately returns `nil` instead.
   ///
   /// The optimizer can often eliminate bounds- and uniqueness-checking
   /// within an algorithm. When that fails, however, invoking the same
-  /// algorithm on `body`\ 's argument may let you trade safety for speed.
+  /// algorithm on the `buffer` argument may let you trade safety for speed.
+  ///
+  /// Your closure should always perform any necessary cleanup, because the
+  /// method makes no guarantees about the state of the collection if the
+  /// closure throws an error. Your changes to the collection may be absent
+  /// from the collection after throwing the error, because the closure could
+  /// receive a temporary copy rather than direct access to the collection's
+  /// storage.
+  ///
+  /// - Warning: Your `body` closure must not replace `buffer`.
+  ///
+  /// Successive calls to this method may provide a different pointer on each
+  /// call. Don't store `buffer` outside of this method.
   ///
   /// A `Collection` that provides its own implementation of this method
   /// must provide contiguous storage to its elements in the same order
-  /// as they appear in the collection. This guarantees that contiguous
-  /// mutable storage to any of its subsequences can be generated by slicing
+  /// as they appear in the collection. This guarantees that it's possible to
+  /// generate contiguous mutable storage to any of its subsequences by slicing
   /// `buffer` with a range formed from the distances to the subsequence's
   /// `startIndex` and `endIndex`, respectively.
   ///
-  /// - Note: `buffer` must not be replaced by `body`.
-  ///
   /// - Parameters:
-  ///   - body: a closure to be executed using the elements of this collection.
-  ///   - buffer: a buffer to the mutable contiguous storage of this collection.
-  /// - Returns: the value returned by `body`, or `nil`.
+  ///   - body: A closure that receives an in-out
+  ///     `UnsafeMutableBufferPointer` to the collection's contiguous storage.
+  ///     Use the buffer pointer directly; don't replace it, because this leads
+  ///     to a crash in all implementations of this method within the standard
+  ///     library.
```

Comment:
The warning note above and the parameter here are sort of splitting up the information.  Is there a way to put both facts in the same place?

Reply:
That's a good point. They're both saying the same thing, a result of combining the text from my old PR (which put the warning in the parameter, as it is here), and the `- Note` from Guillaume's version. We really only need one, so I guess it's a choice of which y'all (@amartini51, @glessard) would prefer:

- The prominence of the warning box, because the consequences of getting it wrong are really bad.
- The guidance here in the parameter, where it specifically relates to the thing you're not allowed to replace.

My preference would be for the second, just because I don't think that many developers are going to even think to replace the pointer.

Reply:
I personally prefer the warning. We fairly frequently see failed attempts at replacing the buffer, and my Note was an attempt to make the issue more obvious. I didn't know Warning existed; I might have used that!